shtab/__init__.py

```diff
@@ -275,85 +399,133 @@ def complete_bash(
-  local args_gen="${flags_list}_COMPGEN"
-  [ -n "${!args_gen}" ] && local opts_more="$(${!args_gen} "$word")"
-  local opts="${!flags_list}"
-  if [ -z "$opts$opts_more" ]; then
-    {root_prefix}_compgen_command_ $1
+# $1=action identifier
+# $2=is positional action (boolean)
+# This function is called when a new action is encountered
+# to set all the identifiers for that action's parameters.
+_set_new_action() {
+  current_action="${prefix}_$(_shtab_replace_nonword $1)"
+
+  local current_action_compgen_var=${current_action}_COMPGEN
+  current_action_compgen="${!current_action_compgen_var}"
+
+  local current_action_choices_var="${current_action}_choices"
+  current_action_choices="${!current_action_choices_var}"
+
+  local current_action_nargs_var="${current_action}_nargs"
+  if [ -n "${!current_action_nargs_var}" ]; then
+    current_action_nargs="${!current_action_nargs_var}"
   else
-    COMPREPLY=( $(compgen -W "$opts" -- "$word"; \
-[ -n "$opts_more" ] && echo "$opts_more") )
+    current_action_nargs=1
   fi
+
+  current_action_index=$word_index
+  current_action_is_positional=$2
 }
 
 # Notes:
 # `COMPREPLY` contains what will be rendered after completion is triggered
-# `word` refers to the current typed word
+# `completing_word` refers to the currently typed word to generate completions for
 # `${!var}` is to evaluate the content of `var`
 # and expand its content as a variable
 #       hello="world"
 #       x="hello"
 #       ${!x} ->  ${hello} ->  "world"
 {root_prefix}() {
-  local word="${COMP_WORDS[COMP_CWORD]}"
-
+  local completing_word="${COMP_WORDS[COMP_CWORD]}"
   COMPREPLY=()
 
-  if [ "${COMP_CWORD}" -eq 1 ]; then
-    {root_prefix}_compgen_root_ ${COMP_WORDS[1]}
-  elif [ "${COMP_CWORD}" -eq 2 ]; then
-    {root_prefix}_compgen_command_ ${COMP_WORDS[1]}
-  elif [ "${COMP_CWORD}" -ge 3 ]; then
-    {root_prefix}_compgen_subcommand_ ${COMP_WORDS[1]} ${COMP_WORDS[2]}
+  prefix={root_prefix}
+  word_index=1
+  _set_parser_defaults
+
+  # determined what arguments are appropriate for the current state
+  # of the arg parser
+  while [ $word_index -ne $COMP_CWORD ]
+  do
+
+    local this_word="${COMP_WORDS[$word_index]}"
+
+    # If we encounter a valid subcommand,
+    # add it to the prefix and reset the current action
+    if [[ -n $subparsers && "${subparsers[@]}" =~ "${this_word}" ]]; then
+      prefix="${prefix}_$(_shtab_replace_nonword $this_word)"
+      _set_parser_defaults
+    fi
+
+    # Check to see if a new action should be acquired,
+    # either because an option string is recognized,
+    # or because no more input is expected from the current action,
+    # indicating that the next positional action can fill in here.
+    if [[ "${current_option_strings[@]}" =~ "${this_word}" ]]; then
+      _set_new_action $this_word false
+      optional_action_offset=0
+    else
+      optional_action_offset=1
+    fi
+
+    if [[ "$current_action_nargs" != "*" ]] && \\
```

Comment:
```suggestion
    if [[ "$current_action_nargs" != "?" ]] && \\
       [[ "$current_action_nargs" != "*" ]] && \\
       [[ "$current_action_nargs" != "+" ]] && \\
```

should there be an `else` clause too though?

Reply:
Got it.  The `?` case is actually accounted for already, as it is parsed earlier in the python script to the default "nargs = 1" case.  As a result, the `if [[ "$current_action_nargs" != "?" ]] && \\` line is redundant.

I don't think an else clause is necessary here.  The if statement evaluates to true when the number of arguments for a given action has reached nargs.  When this occurs, the script moves on to the next positional action.  In the case where nargs has not yet been reached (or will never be reached, in the case of `*` or `+`) the script only needs to _not_ move on to the next action.  This is equivalent to just not executing the code within the if statement, so no else statement is necessary.  At least, I _think_ that's how it should work.  Please let me know if my logic is faulty here.

Reply:
Hmmm...  I just had another thought:  Do we need to account for the case where arguments to a positional action are split up by an optional action?  Is this even allowed by argparse? 
e.g. `command arg_1_to_pos_action_1 --opt-action arg_to_opt-action arg_2_to_pos_action_1 arg_to_pos_action_2`
Currently, this isn't expected or handled properly by the current script.

Reply:
`argparse` does indeed allow `<pos> --opt <pos>` :)